{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Input geneblock\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations \n",
    "\n",
    "data=pd.read_csv(\"problematic antibodies.csv\")\n",
    "H1=data.loc[25:34]\n",
    "H2=data.loc[49:56]\n",
    "H3=data.loc[96:109]\n",
    "L1=data.loc[144:153]\n",
    "L2=data.loc[171:177]\n",
    "L3=data.loc[210:218]\n",
    "Position=data.iloc[0:121,0]\n",
    "Position_light=data.iloc[122:300,0]\n",
    "\n",
    "# Genblocks bococizumab\n",
    "Genblock_heavy_all=['CAAGTTCAATTGGTTCAATCTGGTGCTGAAGTTAAGAAACCAGGTGCTTCAGTTAAAGTTTCTTGTAAAGCTTCTGGTTATACTTTTACTTCTTATTACATGCATTGGGTTAGACAAGCTCCAGGTCAAGGTTTGGAATGGATGGGTGAAATCTCTCCATTTGGTGGTAGAACTAACTATAATGAAAAGTTTAAGTCTAGAGTTACTATGACTAGAGATACTTCTACATCTACTGTATATATGGAATTGTCTTCTTTGAGATCTGAAGATACTGCTGTTTATTACTGTGCTAGAGAAAGACCATTGTATGCTTCTGATTTGTGGGGTCAGGGTACTACTGTTACTGTTTCTTCT',\n",
    "                    'GAAGTTCAATTAGTTGAATCCGGTGGTGGTTTGGTTCAACCAGGTGGTTCTCTTAGATTGTCTTGCGCTGCTTCTGGTTTTACTTTGTCTGGTGATTGGATTCATTGGGTTAGACAAGCTCCAGGTAAAGGTTTGGAATGGGTTGGTGAAATTAGTGCTGCTGGGGGTTATACTGATTATGCTGATTCTGTTAAAGGTAGATTTACTATTTCTGCTGATACTTCTAAAAACACTGCTTACTTGCAAATGAATTCTTTGAGAGCTGAAGATACTGCTGTCTATTATTGCGCTAGAGAATCTAGAGTTTCTTTTGAAGCTGCTATGGATTATTGGGGTCAAGGTACTTTGGTTACTGTTTCTTCT',\n",
    "                    'CAAGTTCAATTGGTTCAATCTGGTGCTGAAGTTAAAAAGCCAGGTGCTTCTGTTAAAGTTTCTTGTAAAGCTTCTGGTTATTCTTTTACCAACTACTATATTCATTGGGTTAGACAAGCTCCAGGTCAAAGATTGGAATGGATGGGTTGGATTAATGCTGGTAATGGTAATACAAAGTATTCACAAAAATTCCAAGGAAGAGTTACTATTACGAGAGATACTTCTGCTTCTACTGCTTATATGGAGTTGTCTTCTTTGAGATCTGAAGATACTGCTGTATACTATTGTGTTAGGAGACAAAGGTTTCCATATTACTTTGATTATTGGGGTCAAGGTACTTTGGTTACTGTTTCTTCT',\n",
    "                    'CAAGTTCAACTTGTTCAATCTGGTGCTGAAGTTAAGAAACCAGGTGCTTCTGTTAAAGTTTCTTGTAAAGCTTCTGGTTATACTTTTACTGATTACTATATGCATTGGGTTAGACAAGCTCCAGGTCAGGGTTTGGAATGGATGGGTAGAGTTAATCCAAATAGAAGAGGAACTACTTATAATCAAAAGTTTGAAGGTCGTGTAACTATGACTACTGATACTTCTACGTCTACTGCTTATATGGAATTGCGTTCTCTTAGATCTGATGATACTGCTGTATATTACTGTGCTAGAGCTAATTGGTTGGACTATTGGGGTCAAGGCACTACCGTTACTGTTTCTTCT']\n",
    "Genblock_light_all=['GATATTCAAATGACTCAATCTCCATCTTCTTTGTCTGCTTCTGTTGGTGATAGAGTTACTATAACTTGTCGTGCTTCTCAGGGTATTTCTTCTGCGTTGGCTTGGTATCAGCAAAAACCAGGTAAAGCTCCAAAATTGTTGATTTATTCCGCTTCTTATAGATATACTGGTGTTCCATCTCGTTTTTCCGGTTCAGGTTCTGGTACTGATTTTACTTTTACTATTTCTAGTTTGCAACCAGAAGATATTGCTACTTATTACTGTCAACAAAGATACTCTCTTTGGAGAACTTTTGGTCAAGGTACTAAATTGGAAATTAAG',\n",
    "                    'GATATTCAAATGACTCAATCTCCATCATCTTTGTCTGCTTCTGTTGGTGATAGAGTTACTATTACTTGTAGAGCTTCTCAAAATATCGCTACTGATGTTGCTTGGTATCAACAAAAACCAGGTAAAGCTCCAAAACTATTGATCTATTCTGCATCTTTCTTGTATTCTGGTGTTCCAAGTAGATTTTCAGGTAGCGGCTCTGGCACGGACTTCACTTTGACTATTTCTTCTTTGCAACCAGAGGATTTTGCTACATATTACTGTCAGCAATCTGAACCAGAACCATATACTTTTGGTCAAGGTACTAAAGTTGAAATTAAG',\n",
    "                    'GAAATTGTTTTGACTCAGTCTCCGGCCACTTTGTCTGTTTCTCCAGGTGAAAGAGCTACGTTGTCTTGTAGAGCTTCTCAATCTGTTGGGACTAATGTTGCTTGGTATCAACAAAAACCAGGTCAAGCTCCAAGAGTTTTGATATATTCTACTTCTTCTAGAGCCACTGGTATTACTGATAGATTTTCAGGTTCTGGATCTGGTACTGATTTTACTTTGACTATTTCTAGATTGGAACCCGAAGATTTTGCTGTTTACTATTGTCAACAGTTTAACAAATCTCCCTTGACTTTTGGTGGTGGTACTAAAGTTGAAATTAAG',\n",
    "                    'GATATTCAAATGACTCAATCTCCATCTTCTTTGTCTGCTAGTGTTGGTGATAGAGTTACTATTACTTGTTCTGTTTCCTCTTCCGTTTCTTCTATTTACTTGCATTGGTATCAACAAAAACCAGGTAAAGCTCCAAAACTTTTGATTTATTCTACTTCTAATTTGGCTTCTGGAGTTCCCTCTAGATTTTCCGGTTCTGGAAGCGGGACTGATTTTACTTTGACTATTTCTTCCCTACAACCAGAAGATTTTGCTACTTATTACTGCCAAGTTTACTCCGGTTATCCATTGACTTTTGGTGGTGGTACTAAAGTTGAAATTAAG']\n",
    "#len(Genblock)\n",
    "# which antibody\n",
    "# bococizumab:1, duligotuzumab:2, lenzilumab:3, emibutuzumab:4\n",
    "n2=3\n",
    "n1=n2+4\n",
    "Heavy_f1=data.iloc[0:121,n2+4]\n",
    "Light_f1=data.iloc[121:300,n2+4]\n",
    "Genblock=Genblock_heavy_all[n2-1]\n",
    "Genblock_light=Genblock_light_all[n2-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Input definition and constants\n",
    "# generate a codon table to map the amino acid\n",
    "def translate(seq): \n",
    "       \n",
    "    table = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_', \n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W', \n",
    "    } \n",
    "    protein =\"\" \n",
    "    if len(seq)%3 == 0: \n",
    "        for i in range(0, len(seq), 3): \n",
    "            codon = seq[i:i + 3] \n",
    "            protein+= table[codon] \n",
    "    return protein\n",
    "\n",
    "def Symbol3(cod):\n",
    "    \n",
    "    table={'CGT':'B','CTG':'B','GTC':'B','GCT':'B','TCG':'B','TGC':'B',\n",
    "           'AGT':'D','ATG':'D','TAG':'D','TGA':'D','GTA':'D','GAT':'D',\n",
    "           'ACT':'H','ATC':'H','TAC':'H','TCA':'H','CTA':'H','CAT':'H',\n",
    "           'ACG':'V','AGC':'V','CGA':'V','CAG':'V','GAC':'V','GCA':'V'}\n",
    "    protein=\"\"\n",
    "    if len(cod)%3 ==0:\n",
    "        for i in range(0, len(cod), 3): \n",
    "            codon = cod[i:i + 3] \n",
    "            protein+= table[codon] \n",
    "    return protein\n",
    "\n",
    "def Symbol2(cod):\n",
    "    \n",
    "    table={'AT':'W','CG':'S','AC':'M','GT':'K','AG':'R','CT':'Y',\n",
    "          'TA':'W','GC':'S','CA':'M','TG':'K','GA':'R','TC':'Y'}\n",
    "    protein=\"\"\n",
    "    if len(cod)%2 ==0:\n",
    "        for i in range(0, len(cod), 2): \n",
    "            codon = cod[i:i + 2] \n",
    "            protein+= table[codon] \n",
    "    return protein\n",
    "\n",
    "#based on the designed library\n",
    "def DeGCod(AA):\n",
    "    \n",
    "    table={'R':'RVR','Y':'KHY','W':'KBG','A':'KHY','L':'VWA','G':'RVY','P':'SHR','F':'KHY','V':'SHR',\n",
    "          'H':'-','I':'RHA','K':'RHA','M':'-'}\n",
    "    protein=\"\"\n",
    "    if len(AA)%1==0:\n",
    "        for i in range(0,len(AA),1):\n",
    "            codon = AA[i:i+1]\n",
    "            protein+=table[AA]\n",
    "    return protein\n",
    "def DeGCod_comp(AA):\n",
    "    \n",
    "    table={'R':'YBY','Y':'MDR','W':'MVC','A':'MDR','L':'BWT','G':'YBR','P':'SDY','F':'MDR','V':'SDY',\n",
    "          'H':'-','I':'YDT','K':'YDT','M':'-'}\n",
    "    protein=\"\"\n",
    "    if len(AA)%1==0:\n",
    "        for i in range(0,len(AA),1):\n",
    "            codon = AA[i:i+1]\n",
    "            protein+=table[AA]\n",
    "    return protein\n",
    "\n",
    "# generate complementary chain\n",
    "def complement(seq):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'} \n",
    "    bases = list(seq) \n",
    "    bases = [complement[base] for base in bases] \n",
    "    return ''.join(bases)\n",
    "    def reverse_complement(s):\n",
    "        return complement(s[::-1])\n",
    "    \n",
    "# reverse chain from 5'-3'\n",
    "def reverse(s): \n",
    "    if len(s) == 0: \n",
    "        return s \n",
    "    else: \n",
    "        return reverse(s[1:]) + s[0] \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#limit for natural diversity of 50%\n",
    "x=np.array([[\"HCDR1\",\"H26\",\"G\"],\n",
    "           [\"HCDR1\",\"H29\",\"F\"],\n",
    "           [\"HCDR1\",\"H32\",\"Y\"],\n",
    "           [\"HCDR2\",\"H51\",\"I\"],\n",
    "           [\"HCDR2\",\"H55\",\"G\"],\n",
    "           [\"HCDR3\",\"H93\",\"A\"],\n",
    "           [\"HCDR3\",\"H94\",\"R\"]])\n",
    "NDAA=pd.DataFrame(x,columns=['Region','Position','AA'])\n",
    "y=np.array([[\"LCDR1\",\"L32\",\"Y\"],\n",
    "           [\"LCDR1\",\"L33\",\"L\"],\n",
    "           [\"LCDR2\",\"L51\",\"A\"],\n",
    "           [\"LCDR2\",\"L54\",\"R\"]])\n",
    "NDAA_light=pd.DataFrame(y,columns=['Region','Position','AA'])\n",
    "NDAA.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## heavy chain libraries: identify sites and amino aicd --Step1 \n",
    "df_H1=H1.iloc[:,n1]\n",
    "SASA_H1=H1.iloc[:,n2]\n",
    "df_H2=H2.iloc[:,n1]\n",
    "SASA_H2=H2.iloc[:,n2]\n",
    "df_H3=H3.iloc[:,n1]\n",
    "SASA_H3=H3.iloc[:,n2]\n",
    "position=[]\n",
    "s=0.1\n",
    "s=0\n",
    "for i in range(0,len(df_H1)):\n",
    "    if df_H1.iloc[i]==\"G\" or df_H1.iloc[i]==\"A\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"F\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"H\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"I\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"K\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"L\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"M\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"P\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"R\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"V\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"W\" and SASA_H1.iloc[i]>=s or df_H1.iloc[i]==\"Y\" and SASA_H1.iloc[i]>=s:\n",
    "        position.append((\"HCDR1\",H1.iloc[i,0],df_H1.iloc[i],SASA_H1.iloc[i],DeGCod(df_H1.iloc[i]),DeGCod_comp(df_H1.iloc[i])))\n",
    "\n",
    "for i in range(0,len(df_H2)):\n",
    "    if df_H2.iloc[i]==\"G\" or df_H2.iloc[i]==\"A\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"F\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"H\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"I\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"K\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"L\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"M\" and SASA_H2.iloc[i]>s or df_H2.iloc[i]==\"P\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"R\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"V\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"W\" and SASA_H2.iloc[i]>=s or df_H2.iloc[i]==\"Y\" and SASA_H2.iloc[i]>=s:\n",
    "        position.append((\"HCDR2\",H2.iloc[i,0],df_H2.iloc[i],SASA_H2.iloc[i],DeGCod(df_H2.iloc[i]),DeGCod_comp(df_H2.iloc[i])))\n",
    "        \n",
    "for i in range(0,len(df_H3)):\n",
    "    if df_H3.iloc[i]==\"G\" or df_H3.iloc[i]==\"A\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"F\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"H\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"I\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"K\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"L\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"M\" and SASA_H3.iloc[i]>s or df_H3.iloc[i]==\"P\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"R\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"V\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"W\" and SASA_H3.iloc[i]>=s or df_H3.iloc[i]==\"Y\" and SASA_H3.iloc[i]>=s:\n",
    "        position.append((\"HCDR3\",H3.iloc[i,0],df_H3.iloc[i],SASA_H3.iloc[i],DeGCod(df_H3.iloc[i]),DeGCod_comp(df_H3.iloc[i])))\n",
    "        \n",
    "position\n",
    "position_df=pd.DataFrame(position,columns=['Region','Position','Original AA','SASA score','De codon','De codon comp'])  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Heavy chain libraries: eliminate sites with max ND--Step2\n",
    "row=[]\n",
    "for i in range (position_df.shape[0]):\n",
    "    for j in range(NDAA.shape[0]):\n",
    "        if position_df.iloc[i,0]==NDAA.iloc[j,0] and position_df.iloc[i,1]==NDAA.iloc[j,1] and position_df.iloc[i,2]==NDAA.iloc[j,2]:\n",
    "            #print(position_df.iloc[i,1])\n",
    "            row.append(i)\n",
    "position_drop=position_df.drop(row)\n",
    "position_drop.sort_values(by=['Region'], ascending=True)\n",
    "print len(position_drop)\n",
    "print position_drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Heavy chain: change degenerate codon--Step3 \n",
    "mapcod=[]\n",
    "mapcod=pd.DataFrame({'Map codon':[]})\n",
    "mapcod_ori=pd.DataFrame({'Map codon origin':[]})\n",
    "mapcod_c=pd.DataFrame({'Map codon comp':[]})\n",
    "mapcod_c_ori=pd.DataFrame({'Map codon comp origin':[]})\n",
    "for i in range (0,len(Genblock),3):\n",
    "    mapcod=mapcod.append({'Map codon':Genblock[i:i+3]}, ignore_index=True)\n",
    "    mapcod_ori=mapcod_ori.append({'Map codon origin':Genblock[i:i+3]}, ignore_index=True)\n",
    "    mapcod_c=mapcod_c.append({'Map codon comp':complement(Genblock[i:i+3])}, ignore_index=True)\n",
    "    mapcod_c_ori=mapcod_c_ori.append({'Map codon comp origin':complement(Genblock[i:i+3])}, ignore_index=True)\n",
    "    \n",
    "Heavy_f1_modi = pd.concat([Position, Heavy_f1], axis=1, sort=False)\n",
    "Heavy_f1_new=Heavy_f1_modi.drop(Heavy_f1_modi[Heavy_f1_modi.iloc[:,1]=='-'].index,axis=0)\n",
    "Heavy_f1_new2 = pd.DataFrame([Heavy_f1_new.iloc[:,0].values,Heavy_f1_new.iloc[:,1].values]).T\n",
    "Heavy_frame = pd.concat([Heavy_f1_new2,mapcod, mapcod_ori,mapcod_c, mapcod_c_ori], axis=1, sort=False)\n",
    "\n",
    "List=Heavy_frame.iloc[:,0]\n",
    "unique_index = pd.Index(List)\n",
    "for i in range(len(position_drop)):\n",
    "    index=unique_index.get_loc(position_drop.iloc[i,1])\n",
    "    Heavy_frame.iloc[index,2]=position_drop.iloc[i,4] \n",
    "    Heavy_frame.iloc[index,4]=position_drop.iloc[i,5] \n",
    "# calculate primer's starting position and end position\n",
    "unique_index = pd.Index(Heavy_frame.iloc[:,0])\n",
    "l1s=unique_index.get_loc('H26')\n",
    "l1e=unique_index.get_loc('H35')\n",
    "l2s=unique_index.get_loc('H50')\n",
    "l2e=unique_index.get_loc('H65')\n",
    "l3s=unique_index.get_loc('H93')\n",
    "l3e=unique_index.get_loc('H102')\n",
    "print l1s,l1e,l2s,l2e,l3s,l3e\n",
    "#Light_frame.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plan A primer design\n",
    "#forward 3rd primer\n",
    "F3_1f=''\n",
    "F3_1r=''\n",
    "# generate a broader region for H1 region\n",
    "f3_1f=Heavy_frame.iloc[l1s-8:l1e+1+14,2]\n",
    "f3_1r=Heavy_frame.iloc[l1s-8:l1e+1+14,4]\n",
    "for i in range (len(f3_1f)):\n",
    "    F3_1f=F3_1f+f3_1f.iloc[i]\n",
    "    F3_1r=F3_1r+f3_1r.iloc[i]\n",
    "\n",
    "# not paired sequence after H1\n",
    "F3_3f=''\n",
    "F3_3r=''\n",
    "f3_3f=Heavy_frame.iloc[l1e+7:l1e+14,2]\n",
    "f3_3r=Heavy_frame.iloc[l1e+7:l1e+14,4]\n",
    "for i in range (len(f3_3f)):\n",
    "    F3_3f=F3_3f+f3_3f.iloc[i]\n",
    "    F3_3r=F3_3r+f3_3r.iloc[i]\n",
    "\n",
    "# sequence for H1 region\n",
    "F3_2f=''\n",
    "F3_2r=''\n",
    "f3_2f=Heavy_frame.iloc[l1s:l1e+1,2]\n",
    "f3_2r=Heavy_frame.iloc[l1s:l1e+1,4]\n",
    "for i in range (len(f3_2f)):\n",
    "    F3_2f=F3_2f+f3_2f.iloc[i]\n",
    "    F3_2r=F3_2r+f3_2r.iloc[i]\n",
    "    \n",
    "F3_H1startf=F3_1f.find(F3_2f)\n",
    "F3_1f.rfind('G')\n",
    "\n",
    "# 2nd Primer, reverse sequence before H1 region\n",
    "F2_1r=F3_1r[min(F3_1f.index('C'),F3_1f.index('G')):F3_H1startf]\n",
    "Tm_1 = 4*(F2_1r.count('C')+F2_1r.count('G'))+2*(F2_1r.count('A')+F2_1r.count('T'))\n",
    "Tm_2 = 4*(F3_3f.count('C')+F3_3f.count('G'))+2*(F3_3f.count('A')+F3_3f.count('T'))\n",
    "# \n",
    "print Tm_1,Tm_2,len(F3_1f),len(F3_3f)\n",
    "print F3_2f, len(F3_2f)\n",
    "\n",
    "# sequence for H2 region (reverse)\n",
    "F4_2f=''\n",
    "F4_2r=''\n",
    "f4_2f=Heavy_frame.iloc[l2s:l2e+1,2]\n",
    "f4_2r=Heavy_frame.iloc[l2s:l2e+1,4]\n",
    "for i in range (len(f4_2f)):\n",
    "    F4_2f=F4_2f+f4_2f.iloc[i]\n",
    "    F4_2r=F4_2r+f4_2r.iloc[i]\n",
    "    \n",
    "F4_2r=reverse(F4_2r[0:len(F4_2r)])\n",
    "print F4_2r, len(F4_2r)\n",
    "\n",
    "# sequence for H3 region (reverse)\n",
    "F6_2f=''\n",
    "F6_2r=''\n",
    "f6_2f=Heavy_frame.iloc[l3s:l3e+1,2]\n",
    "f6_2r=Heavy_frame.iloc[l3s:l3e+1,4]\n",
    "for i in range (len(f6_2f)):\n",
    "    F6_2f=F6_2f+f6_2f.iloc[i]\n",
    "    F6_2r=F6_2r+f6_2r.iloc[i]\n",
    "    \n",
    "F6_2r=reverse(F6_2r[0:len(F6_2r)])\n",
    "print F6_2r,len(F6_2r)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
